### PR TITLE
patch for read replica

### PR DESCRIPTION
**Pull request by lishenzhi**
routing get to replicas.

### CODE REVIEW

File: cacheops/redis.py [hunk 1]
```diff
@@ -65,8 +66,25 @@ def __setattr__(self, name, value):
...
+        """ Proxy `get` calls to redis replica.
+        """
+        def get(self, *args, **kwargs):
+            try:
+                return redis_replica.get(*args, **kwargs)
+            except redis.ConnectionError:
+                return super(ReplicaProxyRedis, self).get(*args, **kwargs)
+
+    redis_client = ReplicaProxyRedis(**settings.CACHEOPS_REDIS)
+except AttributeError:
```
Comment (nicwolff) [inline cacheops/redis.py]: What are you testing for here? Missing `settings.CACHEOPS_REDIS_REPLICA`?

Comment (nicwolff) [inline cacheops/redis.py]: If so maybe `try:` that line alone, then this `except:`, then `else:` the rest of the function?

File: cacheops/redis.py [hunk 2]
```diff
@@ -65,8 +66,25 @@ def __setattr__(self, name, value):
...
+    redis_replica_conf = settings.CACHEOPS_REDIS_REPLICA
+    redis_replica = redis.StrictRedis(**redis_replica_conf)
+
+    class ReplicaProxyRedis(CacheopsRedis):
+        """ Proxy `get` calls to redis replica.
+        """
+        def get(self, *args, **kwargs):
+            try:
+                return redis_replica.get(*args, **kwargs)
+            except redis.ConnectionError:
```
Comment (nicwolff) [inline cacheops/redis.py:85]: So every `get` would have to time out? Can we use a circuit breaker here? Or, can we discover a new replica IP?

Comment (lishenzhi) [inline cacheops/redis.py:85]: the code has a handle_connection_failure already (in this file). it does not have retry though

File: cacheops/redis.py [hunk 3]
```diff
@@ -71,8 +71,12 @@ def __setattr__(self, name, value):
 
 CacheopsRedis = SafeRedis if settings.CACHEOPS_DEGRADE_ON_FAILURE else redis.StrictRedis
 try:
+    # the conf would look like: "redis://cache-001:6379/1,redis://cache-002:6379/2"
     redis_replica_conf = settings.CACHEOPS_REDIS_REPLICA
-    redis_replica = redis.StrictRedis(**redis_replica_conf)
+    redis_replica_servers = redis_replica_conf.split(',')
+    redis_replicas = []
+    for idx in xrange(len(redis_replica_servers)):
+        redis_replicas[idx] = redis.StrictRedis.from_url(redis_replica_servers[idx])
```
Comment (nicwolff) [inline cacheops/redis.py]: or `redis_replicas = map(redis.StrictRedis.from_url, redis_replica_conf.split(','))`

File: cacheops/redis.py [hunk 4]
```diff
@@ -81,6 +85,8 @@ class ReplicaProxyRedis(CacheopsRedis):
         """
         def get(self, *args, **kwargs):
             try:
+                next_replica_index = random.randint(0, len(redis_replicas))
+                redis_replica = redis_replicas[next_replica_index]
```
Comment (nicwolff) [inline cacheops/redis.py]: or `redis_replica = random.choice(redis_replicas)`

File: cacheops/redis.py [hunk 5]
```diff
@@ -65,8 +66,32 @@ def __setattr__(self, name, value):
         self._setup()
         return setattr(self, name, value)
 
-redis_client = LazyRedis()
 
+#redis_client = LazyRedis()
```
Comment (nicwolff) [inline cacheops/redis.py]: Prefer not to commit commented-out code

**Comment by nicwolff** [comment]
👍 Look OK but some style suggestions, could be more concise using `map` and `random.choice`

**Comment by nicwolff** [comment]
👍 👍 ツ

**Comment by cvitan** [comment]
👍  LGTM

**Pull request merged**